exporter/awsemfexporter/emf_exporter.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsemfexporter
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+	"go.opentelemetry.io/collector/obsreport"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter/translator"
+)
+
+const (
+	defaultForceFlushInterval          = 60 * time.Second
+	defaultForceFlushIntervalInSeconds = 60
+)
+
+type emfExporter struct {
+	svcStructuredLog LogClient
+	config           configmodels.Exporter
+	logger           *zap.Logger
+
+	pusherWG           sync.WaitGroup
+	ForceFlushInterval time.Duration
+	shutdownChan       chan bool
+	retryCnt           int
+
+	token string
+}
+
+// New func creates an EMF Exporter instance with data push callback func
+func New(
+	config configmodels.Exporter,
+	params component.ExporterCreateParams,
+) (component.MetricsExporter, error) {
+	if config == nil {
+		return nil, errors.New("emf exporter config is nil")
+	}
+
+	logger := params.Logger
+	// create AWS session
+	awsConfig, session, err := GetAWSConfigSession(logger, &Conn{}, config.(*Config))
+	if err != nil {
+		return nil, err
+	}
+
+	// create CWLogs client with aws session config
+	svcStructuredLog := NewCloudWatchLogsClient(logger, awsConfig, session)
+
+	emfExporter := &emfExporter{
+		svcStructuredLog:   svcStructuredLog,
+		config:             config,
+		ForceFlushInterval: defaultForceFlushInterval,
+		retryCnt:           *awsConfig.MaxRetries,
+		logger:             logger,
+	}
+	if config.(*Config).ForceFlushInterval > 0 {
```

Comment:
Upstream processors in the collector such as batch_retry should handle retries and queuing. Exporter should not need a timer. Just push any data received immediately. Only thing that needs to be handled is making sure maximum records is not exceeded per put. You will likely need a loop for that. See the X-ray exporter.